### PR TITLE
Rustify

### PR DESCRIPTION
**Pull request by mmstick**
This change will make a handful of situations more idiomatic to Rust. It also fixes most of the Clippy errors.

### CODE REVIEW

File: src/hash.rs [hunk 1]
```diff
@@ -72,11 +72,7 @@ impl ZopfliHash {
         let hpos = pos & ZOPFLI_WINDOW_MASK;
         let mut amount = 0;
 
-        let hash_value = if pos + ZOPFLI_MIN_MATCH <= array.len() {
-            array[pos + ZOPFLI_MIN_MATCH - 1]
-        } else {
-            0
-        };
+        let hash_value = array.iter().nth(pos + ZOPFLI_MIN_MATCH - 1).map_or(0, |value| *value);
```
Comment (shepmaster) [inline src/hash.rs]:
``` rust
let hash_value = array.get(pos + ZOPFLI_MIN_MATCH - 1).cloned().unwrap_or(0);
```

Seems better to me. Thoughts?

File: src/deflate.rs [hunk 2]
```diff
@@ -74,14 +74,10 @@ fn optimize_huffman_for_rle(counts: &mut [usize]) {
...
-                let mut count = (sum + stride / 2) / stride;
-                if count < 1 {
-                    count = 1;
-                }
-                if sum == 0 {
-                    // Don't make an all zeros stride to be upgraded to ones.
-                    count = 0;
-                }
+                let count = {
+                    let count = (sum + stride / 2) / stride;
```
Comment (shepmaster) [inline src/deflate.rs]: Perhaps `count` should only be calculated in the `else` branch?

File: src/blocksplitter.rs [hunk 3]
```diff
@@ -210,25 +204,23 @@ pub fn blocksplit_lz77(options: &Options, lz77: &Lz77Store, maxblocks: usize, sp
...
     let mut lz77splitpoints = Vec::with_capacity(maxblocks);
-
     let mut store = Lz77Store::new();
-
     splitpoints.clear();
 
     /* Unintuitively, Using a simple LZ77 method here instead of lz77_optimal
     results in better blocks. */
-    store.greedy(&mut s, in_data, instart, inend);
+    {
```
Comment (shepmaster) [inline src/blocksplitter.rs:209]: Why the extra block here?

File: src/blocksplitter.rs [hunk 4]
```diff
@@ -157,11 +157,7 @@ pub fn blocksplit_lz77(options: &Options, lz77: &Lz77Store, maxblocks: usize, sp
     let mut lstart = 0;
     let mut lend = lz77.size();
 
-    loop {
-        if maxblocks > 0 && numblocks >= maxblocks {
-          break;
-        }
-
+    while maxblocks > 0 && numblocks >= maxblocks {
```
Comment (shepmaster) [inline src/blocksplitter.rs]: This condition doesn't appear to be correct. Previously, the loop would _stop_ when the condition was true, now it keeps going. The golden master files also have differences, but I'm not sure if it's from this line or not.

File: src/blocksplitter.rs (hunk 3 above)
Comment (mmstick) [inline src/blocksplitter.rs:209]: The state is never used after the method call to greedy.

File: src/blocksplitter.rs (hunk 4 above)
Comment (mmstick) [inline src/blocksplitter.rs]: It is correct. `loop if` should be written as a `while` loop.
**edit:** I see that the conditions are not correct though.

File: src/hash.rs (hunk 1 above)
Comment (mmstick) [inline src/hash.rs]: I will make the change.

File: src/deflate.rs (hunk 2 above)
Comment (mmstick) [inline src/deflate.rs]: I'll replace it with `let count = if sum == 0 { 0 } else { cmp::max((sum + stride / 2) / stride, 1) };`

**Comment by carols10cents** [comment]
Thank you!!!! 😻 😻 😻

**Pull request merged**